### ISSUE TITLE
onBye() method for `OutgoingRequestDelegate` interface

### ISSUE THREAD

**Issue by grxnola** [onsip/SIP.js#778]
Currently I can see no way to react to the receiver hanging up other than hooking into the `SessionState.Terminated` stateChange. Along with the existing methods, setting up a method to call on when the Session generated by the request gets a BYE would be a really useful addition to the library.

I saw there is an onBye() in core/ but personally don't see a way to set that. If a method to do this exists already perhaps it could be documented in the example code?

**Comment by mkemmerz** [comment]
I have the samie issue with `Version 0.15.10`. I am using a workaround to get the BYE event if the receiver hangs up:

```
    ua.transport.on('message', function (message) {
        if (message.startsWith('BYE')) {
            // do something
        }
```

I would prefer if a hangup would create an onBye event for the session but this solution works for now.

**Comment by john-e-riordan** [comment]
@grxnola hooking into the `SessionState.Terminated` stateChange is the correct approach to react to the Session terminating - it covers a hangup occurring (BYE sent or received) as well as any other reason the Session might terminate. It allows you to reliably handle session termination.

Is there a reason this is insufficient and/or can you explain the use case for need to handle an incoming BYE? While I can appreciate adding `onBye()` to the delegate for the sake of completeness (if for nothing else), I'm concerned it may create confusion as providing a handler would not be sufficient to detect when a Session ends.

**Comment by inoric** [comment]
Maybe slightly off-topic here, but I also use the hacky ua.transport.on('message', ...) approach for getting the BYE message. In new versions of SIP.js this emits a warning that it's deprecated.

Our use case is getting the Q.850 hangup cause header from the BYE message. Sometimes we make calls using Originate from asterisk, as method of clicking a number to call that works on both physical phones, softphones and webrtc. When using Originate asterisk first sends an INVITE to the extension dialing the number, so it's an incoming instead of an outgoing call, and there is no Inviter with it's requestDelegate that has onReject which contains the reason. But we can currently get it by listening for raw messages and getting the Q.850 hangup cause.

I realise this is not really a standard use case, but it would be great to have the power to do this if needed. I'm not sure how I would upgrade if transport.on('message', ...) is removed.

**Comment by john-e-riordan** [comment]
Getting the Reason header from the BYE seems like reason enough to add onBye() to session delegate.

If an onBye handler is provided, the handler it will be responsible for accepting the BYE (sending the 200 Ok to the BYE). So minimum handler will look something like...

```
session.delegate = {
  onBye: (bye) void => {
    bye.accept();
  }
};
```

**Comment by inoric** [comment]
That would be really nice.
I imagine this would work like the onByeRequest method in Session and parameter bye would be of type IncomingByeRequest, allowing us to access it's .message property, and read the headers?

We actually do use session state for detecting termination and other state transitions, info from the parsed message is just stored in a variable first, so we can show the cause.

**Comment by inoric** [comment]
Oh you already made a pull request.
Looks perfect, for my use case at least.
Thanks for considering this.

**Comment by grxnola** [comment]
Similar use case to inoric. Cheers!

**Comment by james-criscuolo** [comment]
This has been merged and release with `0.15.11`, closing. Thanks again!
